src/Microsoft.AspNetCore.Blazor/Components/ComponentFactory.cs

```diff
@@ -0,0 +1,104 @@
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Blazor.Components
+{
+    internal class ComponentFactory
+    {
+        private readonly IServiceProvider _serviceProvider;
+        private readonly static BindingFlags _injectablePropertyBindingFlags
+            = BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic;
+        private readonly IDictionary<Type, Action<IComponent>> _cachedInitializers
+            = new Dictionary<Type, Action<IComponent>>();
+
+        public ComponentFactory(IServiceProvider serviceProvider)
+        {
+            _serviceProvider = serviceProvider
+                ?? throw new ArgumentNullException(nameof(serviceProvider));
+        }
+
+        public IComponent InstantiateComponent(Type componentType)
+        {
+            if (!typeof(IComponent).IsAssignableFrom(componentType))
+            {
+                throw new ArgumentException($"The type {componentType.FullName} does not " +
+                    $"implement {nameof(IComponent)}.", nameof(componentType));
+            }
+
+            var instance = (IComponent)Activator.CreateInstance(componentType);
+            PerformPropertyInjection(instance);
+            return instance;
+        }
+
+        private void PerformPropertyInjection(IComponent instance)
+        {
+            var instanceType = instance.GetType();
+            if (!_cachedInitializers.TryGetValue(instanceType, out var initializer))
+            {
+                initializer = CreateInitializer(instanceType);
+                _cachedInitializers[instanceType] = initializer;
+            }
+
+            initializer(instance);
+        }
+
+        private Action<IComponent> CreateInitializer(Type type)
+        {
+            // Do all the reflection up front
+            var injectableProperties = type.GetTypeInfo()
+                .GetProperties(_injectablePropertyBindingFlags)
+                .Where(p => p.GetCustomAttribute<InjectAttribute>() != null)
+                .Where(p => p.SetMethod != null);
+            var injectables = injectableProperties.Select(property =>
+            (
+                propertyName: property.Name,
+                propertyType: property.PropertyType,
+                setter: (IPropertySetter)Activator.CreateInstance(
+                    typeof(PropertySetter<,>).MakeGenericType(type, property.PropertyType),
+                    property.SetMethod)
+            )).ToArray();
+
+            // Return an action whose closure can write all the injected properties
+            // without any further reflection calls (just typecasts)
+            return instance =>
+            {
+                foreach (var injectable in injectables)
+                {
+                    var serviceInstance = _serviceProvider.GetService(injectable.propertyType);
```

Comment:
Any reason not to just reuse https://github.com/aspnet/DependencyInjection/blob/88c3bd6fe2786dd759b4a6c6d7c410e895336b6c/src/DI.Abstractions/ServiceProviderServiceExtensions.cs#L38

Reply:
I was just trying to make the error message clearer (i.e., by specifying which property on which type we failed to bind, rather than only saying which service type isn't registered).

Reply:
OK cool. I think that's a good thing